data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/S3Sink.java

```diff
@@ -82,13 +86,27 @@ public S3Sink(final PluginSetting pluginSetting,
         bufferFactory = new CompressionBufferFactory(innerBufferFactory, compressionEngine, codec);
 
         ExtensionProvider extensionProvider = StandardExtensionProvider.create(codec, compressionOption);
-        KeyGenerator keyGenerator = new KeyGenerator(s3SinkConfig, extensionProvider);
+        KeyGenerator keyGenerator = new KeyGenerator(s3SinkConfig, extensionProvider, expressionEvaluator);
+
+        if (s3SinkConfig.getObjectKeyOptions().getPathPrefix() != null &&
+            !expressionEvaluator.isValidFormatExpression(s3SinkConfig.getObjectKeyOptions().getPathPrefix())) {
+            throw new InvalidPluginConfigurationException("path_prefix is not a valid format expression");
+        }
+
+        if (s3SinkConfig.getObjectKeyOptions().getNamePattern() != null &&
+                !expressionEvaluator.isValidFormatExpression(s3SinkConfig.getObjectKeyOptions().getNamePattern())) {
+            throw new InvalidPluginConfigurationException("name_pattern is not a valid format expression");
+        }
 
         S3OutputCodecContext s3OutputCodecContext = new S3OutputCodecContext(OutputCodecContext.fromSinkContext(sinkContext), compressionOption);
 
         codec.validateAgainstCodecContext(s3OutputCodecContext);
 
-        s3SinkService = new S3SinkService(s3SinkConfig, bufferFactory, codec, s3OutputCodecContext, s3Client, keyGenerator, Duration.ofSeconds(5), pluginMetrics);
+        final S3GroupIdentifierFactory s3GroupIdentifierFactory = new S3GroupIdentifierFactory(keyGenerator);
+        final S3GroupManager s3GroupManager = new S3GroupManager(s3SinkConfig, s3GroupIdentifierFactory, bufferFactory, s3Client);
+
+
+        s3SinkService = new S3SinkService(s3SinkConfig, codec, s3OutputCodecContext, s3Client, keyGenerator, Duration.ofSeconds(5), pluginMetrics, s3GroupManager);
```

Comment:
nit: what is the duration ? Can you move that to constant ?

Reply:
This was already here. Looks like the variable is called `retrySleepTime`. I can make it a constant